### DISCUSSION TITLE
zfsbootmenu on ubuntu 23.04

### DISCUSSION THREAD

**Discussion by evilgausus** [Sithuk/ubuntu-server-zfsbootmenu#29] [General]
Hi,

It looks like upgrading to ubuntu 23.04 breaks zfsbootmenu:

Do you want to continue? [Y/n]
Traceback (most recent call last):
  File "/usr/local/bin/pyznap", line 5, in <module>
    from pyznap.main import main
ModuleNotFoundError: No module named 'pyznap'
E: Problem executing scripts DPkg::Pre-Invoke 'if [ -x /usr/local/bin/pyznap ]; then /usr/local/bin/pyznap snap; fi'
E: Sub-process returned an error code

Has anyone encountered such an issue?

**Comment by Sithuk** [comment]
The script hasn’t been tested to work with anything beyond 22.04. But the error mentions pyznap so it looks like the problem is with pyznap and not zfsbootmenu. You might need to reinstall pyznap. You should be able to use the commands as is from the pyznapinstall function of the script.

How did you upgrade?

**Comment by mishac** [comment]
I was able to get it to work after an upgrade to 23.04 with the following steps:

1. comment out (ie add `#` before every line) `/etc/apt/apt.conf.d/80-zfs-snapshot`
2. run `apt install python3-virtualenv`
3. run the following:
  ```
  cd /opt/pyznap
  virtualenv venv
  source venv/bin/activate ##enter virtual env
  pip install pyznap
  deactivate ##exit virtual env
```
4. uncomment `/etc/apt/apt.conf.d/80-zfs-snapshot`

@Sithuk maybe python3-virtualenv should be made a prequisite or something?

**Comment by mishac** [comment]
https://peps.python.org/pep-0668/ it looks like the issue is related to a change in Python 3.11 where it now bitches and moans about "externally managed" packages installed via pip.

**Comment by Sithuk** [comment]
Hi mishac,

I picked up on that python issue a few weeks back during testing and addressed it in:
https://github.com/Sithuk/ubuntu-server-zfsbootmenu/commit/69a5d4e1acd2aed5897c52edf237bb0f938788c0

New installs shouldn't be affected. However, I can see that those who used a previous version of the script that didn't install the following packages might be.
python3-virtualenv
python3-virtualenvwrapper

As you know, the major issue with 23.04 at the moment seems to be that the remoteaccess function doesn't work, as per your issue: 
https://github.com/Sithuk/ubuntu-server-zfsbootmenu/issues/18
I've had a quick look and upstream dracut doesn't seem to have been patched yet.

**Comment by Sithuk** [comment]
So it looks like I solved the issue with remoteaccess. I've updated issue #18 with the solution. The fix should make the script compatible with 23.04 (and 22.10).

**Reply by Sithuk** [reply]
I've rolled out the fixes to the script and tested with both the server and xubuntu 23.04 variants. Everything seems to work.
